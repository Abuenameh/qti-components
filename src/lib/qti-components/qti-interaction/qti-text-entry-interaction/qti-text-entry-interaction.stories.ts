import { html } from 'lit';
import { action } from '@storybook/addon-actions';

import { QtiTextEntryInteraction } from './qti-text-entry-interaction';
import './qti-text-entry-interaction';

const inputWidthClass = [
  '',
  'qti-input-width-2',
  'qti-input-width-1',
  'qti-input-width-3',
  'qti-input-width-4',
  'qti-input-width-6',
  'qti-input-width-10',
  'qti-input-width-15',
  'qti-input-width-20',
  'qti-input-width-72'
];

export default {
  component: 'qti-text-entry-interaction',
  argTypes: {
    // response: { type: 'string' },
    // expectedLength: { type: 'number' },
    // readonly: { description: 'attr: qti-inline-choice-interaction', type: 'boolean' },
    // disabled: { description: 'attr: qti-inline-choice-interaction', type: 'boolean' },
  }
};

export const Interaction = {
  render: args =>
    html`
      <qti-text-entry-interaction
        @qti-register-interaction="${action(`qti-register-interaction`)}"
        @qti-interaction-response="${action(`qti-interaction-response`)}"
        .response=${args.response}
        ?disabled=${args.disabled}
        ?readonly=${args.readonly}
        placeholder-text="totaal"
        class="text-entry-interaction ${inputWidthClass}"
        expected-length=${args.expectedLength}
        pattern-mask=${args.patternMask}
        data-patternmask-message=${args.dataPatternmaskMessage}
        response-identifier="RESPONSE"
      >
      </qti-text-entry-interaction>
    `,

  args: {}
};
