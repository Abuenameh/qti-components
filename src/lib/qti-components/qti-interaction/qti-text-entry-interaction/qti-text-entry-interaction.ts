import { css, html } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { Interaction } from '../internal/interaction/interaction';
import { resetCss } from '../../utilities/reset-styles/reset-shadowroot-styles';
import { customElement, property, state } from 'lit/decorators.js';

@customElement('qti-text-entry-interaction')
export class QtiTextEntryInteraction extends Interaction {
  @property({ type: Number, attribute: 'expected-length' }) expectedLength: number;

  @property({ type: String, attribute: 'pattern-mask' }) patternMask: string;

  @property({ type: String, attribute: 'placeholder-text' }) placeholderText: string;

  @state()
  private _value = '';

  @property({ type: String, attribute: 'class' }) classNames;
  // @watch('classNames', { waitUntilFirstUpdate: true })
  // handleclassNamesChange(old, disabled: boolean) {
  //   const classNames = this.classNames.split(' ');
  //   classNames.forEach((className: string) => {
  //     if (className.startsWith('qti-height-lines')) {
  //       const nrRows = className.replace('qti-height-lines-', '');
  //       if (this.textareaRef) {
  //         this.textareaRef.value.rows = parseInt(nrRows);
  //       }
  //     }
  //   });
  // }

  public set response(value: string | undefined) {
    this._value = value !== undefined ? value : '';
  }

  public validate() {
    return this._value !== '';
  }

  static override get styles() {
    return [
      resetCss,
      css`
        /* PK: display host as block, else design will be collapsed */
        :host {
          display: inline-block;
        }
        input {
          padding: var(--qti-padding, 0.5rem);
        }
      `
    ];
  }

  override render() {
    return html` <input
      part="input"
      spellcheck="false"
      autocomplete="off"
      @keydown="${event => event.stopImmediatePropagation()}"
      @keyup="${this.textChanged}"
      @change="${this.textChanged}"
      type="text"
      placeholder="${ifDefined(this.placeholderText ? this.placeholderText : undefined)}"
      .value="${this._value}"
      size="${ifDefined(this.expectedLength ? this.expectedLength : undefined)}"
      pattern="${ifDefined(this.patternMask ? this.patternMask : undefined)}"
      ?disabled="${this.disabled}"
      ?readonly="${this.readonly}"
    />`;
  }
  // maxlength="${ifDefined(this.expectedLength ? this.expectedLength : undefined)}"

  protected textChanged(event: Event) {
    if (this.disabled || this.readonly) return;
    const input = event.target as HTMLInputElement;
    this.setEmptyAttribute(input.value);
    if (this._value !== input.value) {
      this._value = input.value;
      this.saveResponse(input.value);
    }
  }

  reset(): void {
    this._value = '';
  }

  private setEmptyAttribute(text: string) {
    this.setAttribute('empty', text === '' ? 'true' : 'false');
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-text-entry-interaction': QtiTextEntryInteraction;
  }
}
