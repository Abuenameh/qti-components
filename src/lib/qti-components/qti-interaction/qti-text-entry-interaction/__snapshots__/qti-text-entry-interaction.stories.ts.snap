// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-interaction/qti-text-entry-interaction Default smoke-test 1`] = `
<div id="root-inner">
  <qti-text-entry-interaction placeholder-text="totaal"
                              response-identifier="RESPONSE"
                              class="text-entry-interaction ,qti-input-width-1,qti-input-width-2,qti-input-width-3,qti-input-width-4,qti-input-width-6,qti-input-width-10,qti-input-width-15,qti-input-width-20,qti-input-width-72"
                              expected-length
                              pattern-mask
                              data-patternmask-message
  >
  </qti-text-entry-interaction>
</div>
`;

exports[`lib/qti-components/qti-interaction/qti-text-entry-interaction Formula smoke-test 1`] = `
<div id="root-inner">
  <p>
    Hoeveel kilo CO2 wordt jaarlijks bespaard door 8 zonnepanelen?
  </p>
  <ul>
    <li>
      Levensduur zonnepaneel 25 jaar
    </li>
    <li>
      CO2 uitstoot voor de productie van 1 zonnepaneel is 500 kilo
    </li>
    <li>
      1 zonnepaneel voorkomt 120 kilo CO2 uitstoot per jaar
    </li>
  </ul>
  <p>
    Reken uit:
  </p>
  <p>
    CO2 uitstoot voor productie van de zonnepanelen per jaar is 8 x
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    /
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    =
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    kilo
  </p>
  <p>
    CO2 uitstoot die wordt bespaard per jaar is 8 x
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    -
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    =
    <qti-text-entry-interaction class="qti-input-width-3"
                                response-identifier="RESPONSE"
    >
    </qti-text-entry-interaction>
    kilo
  </p>
</div>
`;

exports[`lib/qti-components/qti-interaction/qti-text-entry-interaction Sizes smoke-test 1`] = `
<div id="root-inner">
  <p>
    qti-input-width-1 :
    <qti-text-entry-interaction class="qti-input-width-1"
                                expected-length="15"
                                response-identifier="RESPONSE1"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-2 :
    <qti-text-entry-interaction class="qti-input-width-2"
                                expected-length="15"
                                response-identifier="RESPONSE2"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-3 :
    <qti-text-entry-interaction class="qti-input-width-3"
                                expected-length="15"
                                response-identifier="RESPONSE3"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-4 :
    <qti-text-entry-interaction class="qti-input-width-4"
                                expected-length="15"
                                response-identifier="RESPONSE4"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-6 :
    <qti-text-entry-interaction class="qti-input-width-6"
                                expected-length="15"
                                response-identifier="RESPONSE5"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-10:
    <qti-text-entry-interaction class="qti-input-width-10"
                                expected-length="15"
                                response-identifier="RESPONSE6"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-15:
    <qti-text-entry-interaction class="qti-input-width-15"
                                expected-length="15"
                                response-identifier="RESPONSE7"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-20:
    <qti-text-entry-interaction class="qti-input-width-20"
                                expected-length="20"
                                response-identifier="RESPONSE8"
    >
    </qti-text-entry-interaction>
  </p>
  <p>
    qti-input-width-72:
    <qti-text-entry-interaction class="qti-input-width-72"
                                expected-length="72"
                                response-identifier="RESPONSE9"
    >
    </qti-text-entry-interaction>
  </p>
</div>
`;
